### PR TITLE
Improve error checking and logic

### PR DESCRIPTION
**Pull request by friendly-bits**
(Updated)
- Implement much more error checking
- Implement hierarchical error propagation
-  Move commands initialization code to init_command() and call that function from all main command functions
- Call abl_luci_exit() before exiting (if it's defined by external sourcing script)
- Kill luci.adblock-lean when stopping or restarting
- Get rid of some of the nested logic
- Collect errors in ${failure_msg} and ${luci_errors} and call reg_failure() once before exiting

### CODE REVIEW

File: adblock-lean [hunk 1]
```diff
@@ -8,6 +8,12 @@
 # Authors: @Lynx and @Wizballs (OpenWrt forum)
 # Contributors: @antonk; @dave14305 (OpenWrt forum)
 
+# global return codes:
+# 0 - Ok
```
Comment (lynxthecat) [inline adblock-lean:12]:
```suggestion
# 0 - OK
```

Comment (lynxthecat) [reply adblock-lean:12]: Or `Success`?

File: adblock-lean [hunk 2]
```diff
@@ -927,49 +996,55 @@ check_for_updates()
...
 	fi
 	rm -f /var/run/adblock-lean/adblock-lean.latest /var/run/adblock-lean/uclient-fetch_err
 }
 
+# args:
 # 1 - (optional) -f to skip check for existing lock
 # 1/2 - action to write to the pid file
+#
+# return codes:
+# 0 - sucess
```
Comment (lynxthecat) [inline adblock-lean:1018]:
```suggestion
# 0 - success
```

**Comment by friendly-bits** [comment]
^ Fixed `sucess`, and `Ok`, Ok? :)

**Comment by friendly-bits** [comment]
^
- manually call status() on 'status' command; call cleanup_and_exit() after status() returns; if sourced then return the correct return code to the sourcing script
(I know that status doesn't need any cleanup but it was more convenient and shorter to call an already existing function, and that function doesn't delete things when not needed)
- log_failure(): register most recent error in `luci_error` variable

**Comment by friendly-bits** [comment]
^
- call `exit` rather than `return` in the code of the main commands
- cleanup_and_exit(): always exit rather than returning
- trap: always exit rather than returning
- move commands initialization code to init_command()
- main command functions: call init_command()
- init_command(): set traps unconditionally
- cleanup_and_exit(): rely on abl_luci_exit() function being defined in sourcing script, rather than on ${abl_sourced} var
- cleanup_and_exit(): call abl_luci_exit() before exit (if defined)

- stop(): accept exit code for argument
- kill_abl_pids(): also kill luci.adblock-lean if it's running

- change var name 'luci_error' -> 'luci_errors'
- log_failure(): collect errors in ${failure_msg} and ${luci_errors} rather than executing report_failure for every error

**Comment by friendly-bits** [comment]
^
- cleanup_and_exit(): check if failure message is set before executing ${report_failure}
- get_stats(): support the '-noexit' argument to return to the calling function
- status(): call gen_stats() with the '-noexit argument'
- rename log_failure() to reg_failure()

File: adblock-lean [hunk 3]
```diff
@@ -140,21 +151,52 @@ log_msg()
...
+		1) reg_failure "Error: failed to export the blocklist."; return 1 ;;
+		2) return 2
+	esac
+	:	
+}
+
+# return codes:
+# 0 - blocklist file present
+# 1 - blocklist file not present
+check_active_blocklist_file()
```
Comment (lynxthecat) [inline adblock-lean:174]: Nice sequence!

Comment (friendly-bits) [reply adblock-lean:174]: You mean the code comments?

**Comment by friendly-bits** [comment]
^
- bytes2human(): fix the comment
- stop(): support optional '-noexit' argument, return to caller when specified
- restart(): implement function (otherwise restart wouldn't work)
- reload(): implement function stub - simply calls restart()
- update_pid_action(): accept rc=3 from check_lock(), reject rc=2
- check_lock(): check if current PID owns the lock file, and if so then return code 3

The above is mainly to fix boot, restart and reload commands

**Comment by friendly-bits** [comment]
^ init_command(): if failed to create/acquire lock, unset ${lock_req}, ${cleanup_req} before exit
(turns out there was a bug where running the start command while another command was running would remove the lock and do cleanup, which should not happen)

**Comment by lynxthecat** [comment]
I think safe to merge this in now?

**Pull request merged**